projects/utils-maps/src/lib/ol/utils-ol-mapstate.ts

```diff
@@ -0,0 +1,204 @@
+  const view = map.getView();
+  if (!view) {
+    // the map does not have a view, so we can't act
+    // upon it
+    return;
+  }
+  const delta = 1, duration = 250;
+  const currentZoom = view.getZoom();
+  if (currentZoom !== undefined) {
+    const newZoom = view.getConstrainedZoom(currentZoom + delta);
+    if (duration > 0) {
+      if (view.getAnimating()) {
+        view.cancelAnimations();
+      }
+      view.animate({
+        zoom: newZoom,
+        duration,
+        easing: easeOut
+      });
+    } else {
+      view.setZoom(newZoom);
+    }
+  }
+}
+
+
+
+export function getMapProjection(map: Map) {
+  return map.getView().getProjection();
+}
+
+
+/**
+ * vector layers will be reprojected automatically
+ * wms layers will be updated with corresponding proj def in the requests.
+ * for other raster layers and for those wms layers whose backend does not support target projection, please
+ * define initial(default) layer projection, so openlayers will reproject on the client side
+ * projection is proj~ProjectionLike
+ */
+export function setMapProjection(map: Map, projection: Projection | string, viewOptions?: ViewOptions, cb?: (projection: Projection) => void) {
+  let epsg = DEFAULT_PROJECTION;
+  let view = null;
+  if (projection) {
+    let newViewOptions: ViewOptions = {};
+    if (viewOptions) {
+      newViewOptions = viewOptions;
+      newViewOptions.minResolution = undefined;
+      newViewOptions.maxResolution = undefined;
+      newViewOptions.resolution = undefined;
+      newViewOptions.resolutions = undefined;
+    }
+    if (projection instanceof Projection) {
+      newViewOptions.projection = projection;
+      const newCenter = transform(map.getView().getCenter(), map.getView().getProjection(), projection); // get center coordinates in the new projection
+      newViewOptions.center = newCenter; // map.getView().getCenter();
+      // _viewOptions.extent = projection.getExtent();// || undefined;
+      newViewOptions.zoom = map.getView().getZoom();
+    } else if (typeof projection === 'string') {
+      newViewOptions.projection = projection;
+      newViewOptions.center = map.getView().getCenter();
+      newViewOptions.zoom = map.getView().getZoom();
+    }
+    const mapView = new View(newViewOptions);
+    const oldProjection = DEFAULT_PROJECTION;
+    epsg = mapView.getProjection().getCode();
+    map.setView(mapView);
+    view = map.getView();
+
+    // reprojecting vector layers
+    map.getLayers().getArray().forEach((layerGroup: LayerGroup) => {
+      layerGroup.getLayers().getArray().forEach(layer => {
+        if (layer instanceof Layer) {
+          let source = layer.getSource();
+          // check for nested sources, e.g. cluster or cluster of clusters etc
+          while (source['source']) {
+            source = source['source'];
+          }
+          if (source instanceof VectorSource) {
+            reprojectVectorSource(source, oldProjection, DEFAULT_PROJECTION);
```

Comment:
A question to you guys, @boeckMt & @voinSR : could it be that openlayers automatically reprojects vector-sources when the view changes its projection? Or that the reprojection is not necessary because even though a source in projection A, it is being displayed in projection B?

I can do some research on how ol handles vector-sources from different projections, but does anyone of you guys already know that, by any chance?

Reply:
Good news! I did a little research and saw that reprojecting vector-sources *does* make sense. 
Vector-sources do *not* have a `sourceProjection` attribute; only [WMS do](https://openlayers.org/en/latest/doc/tutorials/raster-reprojection.html).

Just for completeness sake, here's a little minimum example showing that vector-data is not reprojected on the fly:

```js
import { Map, View } from 'ol';
import { OSM, Vector as VectorSource } from 'ol/source';
import { Tile as TileLayer, Vector as VectorLayer } from 'ol/layer';
import { GeoJSON } from 'ol/format';
import { get as getProjection } from 'ol/proj';
import 'ol/ol.css';


const container = document.getElementById('mapDiv') as HTMLDivElement;

const button = document.getElementById('toggleProjection') as HTMLButtonElement;

const osm = new TileLayer({
    source: new OSM()
});

const data = new VectorLayer({
    source : new VectorSource({
        features: new GeoJSON().readFeatures({
            "type": "FeatureCollection",
            "features": [{
                "type": "Feature",
                "properties": {},
                "geometry": {
                  "type": "Polygon",
                  "coordinates": [[[11.3818359375,51.767839887322154],[14.23828125,51.767839887322154],[14.23828125,53.05442186546102],[11.3818359375,53.05442186546102],[11.3818359375,51.767839887322154]]]
                }
              }]
          }),
    })
});


const wgs84 = getProjection('EPSG:4326');
const google = getProjection('EPSG:3857');

const googleView = new View({
    center: [1669792, 6800125],
    zoom: 7,
    projection: google
});

const wgs84View = new View({
    center: [15, 52],
    zoom: 7,
    projection: wgs84
});

const views = [wgs84View, googleView];

let activeViewIndex = 0;

const map = new Map({
    view: views[activeViewIndex], 
    layers: [osm, data], 
    target: container
});

button.addEventListener('click', () => {
    activeViewIndex = (activeViewIndex + 1) % 2;
    const newView = views[activeViewIndex];
    map.setView(newView);
});
```